### ISSUE TITLE
Support for Multiple Transformers 1-d then 2-d

### ISSUE THREAD

**Issue by bmweiner** [paulgb/sklearn-pandas#50]
>>>df = pd.DataFrame(['red', 'white','blue','red','blue'], columns=['color'])
    >>>mapper = DataFrameMapper([('color', [preprocessing.LabelEncoder(),
                                            sklearn.preprocessing.OneHotEncoder()])])
    >>>mapper.fit_transform(df)
    array([[ 1.,  1.,  1.,  1.,  1.]])

vs.

    >>>le = sklearn.preprocessing.LabelEncoder()
    >>>enc_color = le.fit_transform(df['color'])
    >>>enc = sklearn.preprocessing.OneHotEncoder()
    >>>enc.fit_transform(enc_color.reshape(-1,1)).toarray()
    array([[ 0.,  1.,  0.],
           [ 0.,  0.,  1.],
           [ 1.,  0.,  0.],
           [ 0.,  1.,  0.],
           [ 1.,  0.,  0.]])

I saw the issue documenting a similar challenge with [sklearn pipelines](https://github.com/scikit-learn/scikit-learn/issues/3956). Is there a way to do this?

**Comment by bmweiner** [comment]
My workaround for single column until OneHotEncoder is updated to accept categorical:

    >>>def make_2d(y):
            return y.reshape(-1,1)

    >>>mapper = DataFrameMapper([
            ('color', [sklearn.preprocessing.LabelEncoder(),
            sklearn.preprocessing.FunctionTransformer(make_2d),
            sklearn.preprocessing.OneHotEncoder()])])
    array([[ 0.,  1.,  0.],
           [ 0.,  0.,  1.],
           [ 1.,  0.,  0.],
           [ 0.,  1.,  0.],
           [ 1.,  0.,  0.]])

**Comment by dukebody** [comment]
I don't think there is an easy obvious way to fix this in `sklearn-pandas`. I believe this is something that is wrong with the `scikit-learn` API, where some transformers accept and output 2-d arrays while some others expect 1-d arrays and return either 1-d arrays or 2-d. This has to be clarified.

Also, as you comment `OneHotEncoder` should be adapted accept non-integer inputs as well, to avoid the need of using a separate label encoding step.

Another workaround for your issue would be to use `LabelBinarizer` instead of `LabelEncoder + OneHotEncoder`.

I'm closing this issue since this is not something I believe we have to fix in `sklearn-pandas`. Please reopen if you don't think the same way. :)

**Comment by bmweiner** [comment]
Thanks @dukebody.
